Os/Generic/PriorityQueue.cpp

```diff
@@ -0,0 +1,186 @@
+    FW_ASSERT(index < this->m_depth);
+    FwSizeType offset = this->m_maxSize * index;
+    ::memcpy(destination, this->m_data + offset, size);
+}
+
+PriorityQueue::~PriorityQueue() {
+    delete[] this->m_handle.m_data;
+    delete[] this->m_handle.m_indices;
+    delete[] this->m_handle.m_sizes;
+}
+
+QueueInterface::Status PriorityQueue::create(const Fw::StringBase& name, FwSizeType depth, FwSizeType messageSize) {
+    // Allocate indices list
+    FwSizeType* indices = new (std::nothrow) FwSizeType[depth];
+    if (indices == nullptr) {
+        return QueueInterface::Status::UNKNOWN_ERROR;
+    }
+    // Allocate sizes list or clean-up
+    FwSizeType* sizes = new (std::nothrow) FwSizeType[depth];
+    if (sizes == nullptr) {
+        delete[] indices;
+        return QueueInterface::Status::UNKNOWN_ERROR;
+    }
+    // Allocate sizes list or clean-up
+    U8* data = new (std::nothrow) U8[depth * messageSize];
+    if (data == nullptr) {
+        delete[] indices;
+        delete[] sizes;
+        return QueueInterface::Status::UNKNOWN_ERROR;
+    }
+    // Allocate max heap or clean-up
+    bool created = this->m_handle.m_heap.create(depth);
+    if (not created) {
+        delete[] indices;
+        delete[] sizes;
+        delete[] data;
+        return QueueInterface::Status::UNKNOWN_ERROR;
+    }
+    // Assign initial indices and sizes
+    for (FwSizeType i = 0; i < depth; i++) {
+        indices[i] = i;
+        sizes[i] = 0;
+    }
+    // Set local tracking variables
+    this->m_handle.m_maxSize = messageSize;
+    this->m_handle.m_indices = indices;
+    this->m_handle.m_data = data;
+    this->m_handle.m_sizes = sizes;
+    this->m_handle.m_startIndex = 0;
+    this->m_handle.m_stopIndex = depth;
+    this->m_handle.m_depth = depth;
+    this->m_handle.m_highMark = 0;
+
+    return QueueInterface::Status::OP_OK;
+}
+
+QueueInterface::Status PriorityQueue::send(const U8* buffer,
+                                           FwSizeType size,
+                                           FwQueuePriorityType priority,
+                                           QueueInterface::BlockingType blockType) {
+    // Check for sizing problem before locking
+    if (size > this->m_handle.m_maxSize) {
+        return QueueInterface::Status::SIZE_MISMATCH;
+    }
+    // Artificial block scope for scope lock ensuring an unlock in all cases and ensuring an unlock before notify
+    {
+        Os::ScopeLock lock(this->m_handle.m_data_lock);
+        if (this->m_handle.m_heap.isFull() and blockType == BlockingType::NONBLOCKING) {
+            return QueueInterface::Status::FULL;
+        }
+        // Will loop and block until full is false
+        while (this->m_handle.m_heap.isFull()) {
+            this->m_handle.m_full.wait(this->m_handle.m_data_lock);
+        }
+        FwSizeType index = this->m_handle.find_index();
+
+        bool pushed = this->m_handle.m_heap.push(priority, index);
+        if (not pushed) {
```

Comment:
may want to assert these conditions instead cause we are protected by the scope lock